File: iEngine-Native/Engine/src/math/Vector2.cpp
#include "../../include/iengine/math/Vector2.h"
#include <cmath>

namespace iengine {

    Vector2::Vector2(float x, float y) : x(x), y(y) {}

    void Vector2::set(float x, float y) {
        this->x = x;
        this->y = y;
    }

    Vector2 Vector2::copy() const {
        return Vector2(x, y);
    }

    Vector2& Vector2::add(const Vector2& v) {
        x += v.x;
        y += v.y;
        return *this;
    }

    Vector2& Vector2::sub(const Vector2& v) {
        x -= v.x;
        y -= v.y;
        return *this;
    }

    Vector2& Vector2::multiplyScalar(float scalar) {
        x *= scalar;
        y *= scalar;
        return *this;
    }

    Vector2& Vector2::divideScalar(float scalar) {
        if (scalar != 0.0f) {
            x /= scalar;
            y /= scalar;
        }
        return *this;
    }

    float Vector2::length() const {
        return std::sqrt(x * x + y * y);
    }

    Vector2& Vector2::normalize() {
        float len = length();
        if (len > 0.0f) {
            x /= len;
            y /= len;
        }
        return *this;
    }

    float Vector2::dot(const Vector2& v) const {
        return x * v.x + y * v.y;
    }

    float Vector2::distanceTo(const Vector2& v) const {
        float dx = x - v.x;
        float dy = y - v.y;
        return std::sqrt(dx * dx + dy * dy);
    }

} // namespace iengine

File: iEngine-Web/src/math/Vector2.ts
export class Vector2 {
    constructor(
        public x: number = 0,
        public y: number = 0
    ) {}
    
    set(x: number, y: number): this {
        this.x = x;
        this.y = y;
        return this;
    }

    toArray(): number[] {
        return [this.x, this.y];
    }
    
    copy(v: Vector2): this {
        this.x = v.x;
        this.y = v.y;
        return this;
    }
    
    add(v: Vector2): this {
        this.x += v.x;
        this.y += v.y;
        return this;
    }
    
    sub(v: Vector2): this {
        this.x -= v.x;
        this.y -= v.y;
        return this;
    }
    
    multiplyScalar(scalar: number): this {
        this.x *= scalar;
        this.y *= scalar;
        return this;
    }
    
    divideScalar(scalar: number): this {
        return this.multiplyScalar(1 / scalar);
    }
    
    length(): number {
        return Math.sqrt(this.x * this.x + this.y * this.y);
    }
    
    normalize(): this {
        return this.divideScalar(this.length() || 1);
    }
    
    dot(v: Vector2): number {
        return this.x * v.x + this.y * v.y;
    }
    
    distanceTo(v: Vector2): number {
        return Math.sqrt(
            (this.x - v.x) ** 2 +
            (this.y - v.y) ** 2
        );
    }
}